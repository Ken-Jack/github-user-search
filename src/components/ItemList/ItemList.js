// @flow

import React from 'react';
import {connect} from 'react-fela';
import map from 'lodash/fp/map';
import Loading from 'components/Loading';

type Props = {
  entities: Object,
  component: ReactClass<*>,
  ids: Array<number>,
  isPending: boolean,
  styles: Object,
};

function ItemList(props: Props) {
  const {
    entities,
    ids,
    isPending,
    styles,
    component: Component,
  } = props;

  if (isPending) {
    return (
      <Loading />
    );
  }

  if (ids.length === 0) {
    return <p>No items to display</p>;
  }

  return (
    <ul>
      {map(id => {
        return (
          <li
            className={styles.ItemList_item}
            key={id}
          >
            <Component {...entities[id]} />
          </li>
        );
      }, ids)}
    </ul>
  );
}

const styles = {
  ItemList_item: () => ({
    borderBottom: '1px solid #ddd',
    marginBottom: '20px',
    paddingBottom: '20px',

    ':last-child': {
      marginBottom: '0',
      paddingBottom: '0',
      borderBottom: '0',
    },
  }),
};

export default connect(styles)(ItemList);
