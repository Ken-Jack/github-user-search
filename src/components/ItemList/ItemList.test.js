import React from 'react';
import felaSnapshot from 'test-util/fela-snapshot';

import ItemList from './ItemList';

function FakeComponent({login}) { // eslint-disable-line
  return <p>{login}</p>;
}

describe('Component: ItemList', () => {

  it('should render components in a list', () => {
    const entities = {
      123: {
        login: 'simon',
      },
      456: {
        login: 'john',
      },
    };
    const component = felaSnapshot(
      <ItemList
        entities={entities}
        ids={[123, 456]}
        isPending={false}
        component={FakeComponent}
      />
    );
    expect(component).toMatchSnapshot();
  });

  it('should display a message when no items exist', () => {
    const component = felaSnapshot(
      <ItemList
        entities={{}}
        ids={[]}
        isPending={false}
        component={FakeComponent}
      />
    );
    expect(component).toMatchSnapshot();
  });

  it('should render a Loading component when isPending', () => {
    const component = felaSnapshot(
      <ItemList
        entities={{}}
        ids={[]}
        isPending={true}
        component={FakeComponent}
      />
    );
    expect(component).toMatchSnapshot();
  });

});
