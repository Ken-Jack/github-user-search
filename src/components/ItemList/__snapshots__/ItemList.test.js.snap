// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: ItemList should display a message when no items exist 1`] = `
Object {
  "component": <p>
    No items to display
  </p>,
  "styles": "
    .a{border-bottom:1px solid #ddd}
    .b{margin-bottom:20px}
    .c{padding-bottom:20px}
    .d:last-child{margin-bottom:0}
    .e:last-child{padding-bottom:0}
    .f:last-child{border-bottom:0}",
}
`;

exports[`Component: ItemList should render a Loading component when isPending 1`] = `
Object {
  "component": <div
    className="u-flex u-flexJustifyCenter u-flexAlignItemsCenter"
  >
    <svg />
  </div>,
  "styles": "
    .a{border-bottom:1px solid #ddd}
    .b{margin-bottom:20px}
    .c{padding-bottom:20px}
    .d:last-child{margin-bottom:0}
    .e:last-child{padding-bottom:0}
    .f:last-child{border-bottom:0}",
}
`;

exports[`Component: ItemList should render components in a list 1`] = `
Object {
  "component": <ul>
    <li
      className="a b c d e f"
    >
      <p>
        simon
      </p>
    </li>
    <li
      className="a b c d e f"
    >
      <p>
        john
      </p>
    </li>
  </ul>,
  "styles": "
    .a{border-bottom:1px solid #ddd}
    .b{margin-bottom:20px}
    .c{padding-bottom:20px}
    .d:last-child{margin-bottom:0}
    .e:last-child{padding-bottom:0}
    .f:last-child{border-bottom:0}",
}
`;
